### PR TITLE
♻️ replace autobind with instance method

### PR DESCRIPTION
**Pull request by michelleyschen**
### WHY are these changes introduced?

Related issue: #991

The `autobind` method from javascript-utilities depend on lodash, which is a heavy library that can slow our performance. This PR is part of the process to get rid of lodash by 
- replacing most of the `autobind` with instance method.
- removing `memoize` because it does not offer any substantial performance enhancement. 

Another PR will follow to handle the cases where I can't directly replace `autobind` with instance method. 

## 

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris-react/blob/master/README.md#development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Versioning%20and%20changelog.md)

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import * as React from 'react';
import {Page} from '@shopify/polaris';

interface State {}

export default class Playground extends React.Component<never, State> {
  render() {
    return (
      <Page title="Playground">
        {/* Add the code you want to test here */}
      </Page>
    );
  }
}
```

</details>

### 🎩 checklist

* [ ] Tested on [mobile](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md#cross-browser-testing)
* [ ] Tested on [multiple browsers](https://help.shopify.com/en/manual/intro-to-shopify/shopify-admin/supported-browsers)
* [ ] Tested for [accessibility](https://github.com/Shopify/polaris-react/blob/master/documentation/Accessibility%20testing.md)

### CODE REVIEW

File: src/components/ResourceList/ResourceList.tsx [hunk 1]
```diff
@@ -111,7 +111,6 @@ export class ResourceList extends React.Component<CombinedProps, State> {
     );
   }
 
-  @autobind
```
Comment (michelleyschen) [inline src/components/ResourceList/ResourceList.tsx]: getter always have the proper scope, we don't need autobind

File: src/components/Autocomplete/components/ComboBox/ComboBox.tsx [hunk 2]
```diff
@@ -132,7 +131,7 @@ export default class ComboBox extends React.PureComponent<Props, State> {
   getChildContext(): Context {
     return {
       comboBoxId: this.state.comboBoxId,
-      selectedOptionId: this.getSelectedOptionId(),
+      selectedOptionId: this.getSelectedOptionId,
```
Comment (AndrewMusgrave) [inline src/components/Autocomplete/components/ComboBox/ComboBox.tsx]: check

File: src/components/Autocomplete/components/ComboBox/ComboBox.tsx [hunk 3]
```diff
@@ -510,29 +491,26 @@ export default class ComboBox extends React.PureComponent<Props, State> {
...
     }
     if (newOption) {
       newOption.active = true;
     }
-  }
+  };
 
-  @autobind
-  private getSelectedOptionId(): string | undefined {
+  private get getSelectedOptionId(): string | undefined {
```
Comment (AndrewMusgrave) [inline src/components/Autocomplete/components/ComboBox/ComboBox.tsx]: check

File: src/components/Autocomplete/components/ComboBox/ComboBox.tsx (hunk 2 above)
Comment (AndrewMusgrave) [inline src/components/Autocomplete/components/ComboBox/ComboBox.tsx]:
```suggestion
      selectedOptionId: this.selectedOptionId,
```

**Comment by michelleyschen** [comment]
i didn't replace all of them mostly because the ones that's left cause un-expected test failure and require more looking into. This PR is already large enough as is so my thought is to get these low risk item out first and high risk item when I have more time.

What do you think?

**Comment by michelleyschen** [comment]
I removed a few more `autobind`, however, this mean I had to replace `debounce` with `lodash/debounce`. The idea will be that we can eventually replace `lodash/dedounce` with our own version as well. but it should be add to the list in #991

**Comment by michelleyschen** [comment]
That's all of `@shopify/javascript-utilities/decorators`!!

**Pull request merged**